pkg/operator/watchdog.go

```diff
@@ -0,0 +1,175 @@
+type DaprWatchdog struct {
+	enabled           bool
+	interval          time.Duration
+	maxRestartsPerMin int
+
+	client         client.Client
+	restartLimiter ratelimit.Limiter
+}
+
+// NeedLeaderElection makes it so the controller runs on the leader node only.
+// Implements sigs.k8s.io/controller-runtime/pkg/manager.LeaderElectionRunnable .
+func (dw *DaprWatchdog) NeedLeaderElection() bool {
+	return true
+}
+
+// Start the controller. This method blocks until the context is canceled.
+// Implements sigs.k8s.io/controller-runtime/pkg/manager.Runnable .
+func (dw *DaprWatchdog) Start(parentCtx context.Context) error {
+	if !dw.enabled {
+		log.Infof("DaprWatchdog is not enabled")
+		return nil
+	}
+
+	log.Infof("DaprWatchdog worker starting")
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+
+	if dw.maxRestartsPerMin > 0 {
+		dw.restartLimiter = ratelimit.New(
+			dw.maxRestartsPerMin,
+			ratelimit.Per(time.Minute),
+			ratelimit.WithoutSlack,
+		)
+	} else {
+		dw.restartLimiter = ratelimit.NewUnlimited()
+	}
+
+	// Use a buffered channel to make sure that there's at most one iteration at a given time
+	// and if an iteration isn't done by the time the next one is scheduled to start, no more iterations are added to the queue
+	workCh := make(chan struct{}, 1)
+	defer close(workCh)
+	firstCompleteCh := make(chan struct{})
+	go func() {
+		defer log.Infof("DaprWatchdog worker stopped")
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-workCh:
+				dw.listPods(ctx)
+				if firstCompleteCh != nil {
+					close(firstCompleteCh)
+					firstCompleteCh = nil
+				}
+			}
+		}
+	}()
+
+	log.Infof("DaprWatchdog worker started")
+
+	// Start an iteration right away, at startup, then wait for completion
+	workCh <- struct{}{}
+	<-firstCompleteCh
+
+	// If we only run once, exit when it's done
+	if dw.interval < time.Second {
+		return nil
+	}
+
+	// Repeat on interval
+	t := time.NewTicker(dw.interval)
+	defer t.Stop()
+
+forloop:
+	for {
+		select {
+		case <-ctx.Done():
+			break forloop
+		case <-t.C:
+			log.Debugf("DaprWatchdog worker tick")
+			select {
+			case workCh <- struct{}{}:
+				// Successfully queued another iteration
+				log.Debugf("Added listPods iteration to the queue")
+			default:
+				// Do nothing - there's already an iteration in the queue
+				log.Debugf("There's already an iteration in the listPods queue-not adding another one")
+			}
+		}
+	}
+
+	log.Infof("DaprWatchdog worker stopping")
+	return nil
+}
+
+func (dw *DaprWatchdog) listPods(ctx context.Context) {
+	log.Infof("DaprWatchdog started checking pods")
+
+	// Request the list of pods
+	// We are not using pagination because we may be deleting pods during the iterations
+	// The client implements some level of caching anyways
+	pod := &corev1.PodList{}
+	err := dw.client.List(ctx, pod,
+		client.InNamespace(GetNamespace()),
```

Comment:
@ItalyPaleAle why only check in the dapr namespace? 🤔

Reply:
Hey good question! I didn't think checking all namespaces was needed but perhaps it is?

Problem would be if you have multiple namespaces with Dapr installed, in which case there may be a race condition.

Perhaps this needs to have another configuration option to pass the namespace(s) to check?

Reply:
There can only be a single Dapr installation in a cluster, but regardless this list operation should list pods in all namespaces.

Reply:
Yes  :) https://github.com/dapr/dapr/pull/4837